Confused error for d.get(k, None) in certain generic context

Example (boiled down from something larger):
```python
from typing import *

_KT = TypeVar('_KT')
_VT = TypeVar('_VT')

def foo(dd):
    # type: (Dict[_KT, _VT]) -> _VT
    for k in dd:
        return dd.get(k, None)
```
The last line gives
```
__tmp__.py:9: error: Argument 1 to "get" of "Mapping" has incompatible type "_KT"; expected "_VT"
```
When I replace `dd.get(k, None)` with `dd.get(k)` the error goes away (even though it should mean the same). This is not new (it behaves the same with version 0.470, and the recent updates to typeshed for get() don't seem to make a difference).

The key factor seems to be the `_VT` context forced upon the `get()` call via the context from the return statement.

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -31,10 +31,10 @@ from mypy import nodes
 from mypy.types import (
     Type, AnyType, CallableType, Void, FunctionLike, Overloaded, TupleType, TypedDictType,
     Instance, NoneTyp, ErrorType, strip_type, TypeType,
-    UnionType, TypeVarId, TypeVarType, PartialType, DeletedType, UninhabitedType,
+    UnionType, TypeVarId, TypeVarType, PartialType, DeletedType, UninhabitedType, TypeVarDef,
     true_only, false_only, function_type
 )
-from mypy.sametypes import is_same_type
+from mypy.sametypes import is_same_type, is_same_types
 from mypy.messages import MessageBuilder
 import mypy.checkexpr
 from mypy.checkmember import map_type_from_supertype, bind_self, erase_to_bound
@@ -2744,9 +2744,10 @@ def is_more_general_arg_prefix(t: FunctionLike, s: FunctionLike) -> bool:
     """Does t have wider arguments than s?"""
     # TODO should an overload with additional items be allowed to be more
     #      general than one with fewer items (or just one item)?
-    # TODO check argument kinds
+    # TODO check argument kinds and otherwise make more general
     if isinstance(t, CallableType):
         if isinstance(s, CallableType):
+            t, s = unify_generic_callables(t, s)
             return all(is_proper_subtype(args, argt)
                        for argt, args in zip(t.arg_types, s.arg_types))
     elif isinstance(t, FunctionLike):
@@ -2755,6 +2756,43 @@ def is_more_general_arg_prefix(t: FunctionLike, s: FunctionLike) -> bool:
                 return all(is_same_arg_prefix(items, itemt)
                            for items, itemt in zip(t.items(), s.items()))
     return False
+
+
+def unify_generic_callables(t: CallableType,
+                            s: CallableType) -> Tuple[CallableType,
+                                                      CallableType]:
+    """Make type variables in generic callables the same if possible.
+
+    Return updated callables. If we can't unify the type variables,
+    return the unmodified arguments.
+    """
+    # TODO: Use this elsewhere when comparing generic callables.
+    if t.is_generic() and s.is_generic():
+        t_substitutions = {}
+        s_substitutions = {}
+        for tv1, tv2 in zip(t.variables, s.variables):
+            # Are these something we can unify?
+            if tv1.id != tv2.id and is_equivalent_type_var_def(tv1, tv2):
+                newdef = TypeVarDef.new_unification_variable(tv2)
+                t_substitutions[tv1.id] = TypeVarType(newdef)
+                s_substitutions[tv2.id] = TypeVarType(newdef)
+        return (cast(CallableType, expand_type(t, t_substitutions)),
+                cast(CallableType, expand_type(s, s_substitutions)))
+    return t, s
+
+
+def is_equivalent_type_var_def(tv1: TypeVarDef, tv2: TypeVarDef) -> bool:
+    """Are type variable definitions equivalent?
+
+    Ignore ids, locations in source file and names.
+    """
+    return (
+        tv1.variance == tv2.variance
+        and is_same_types(tv1.values, tv2.values)
+        and ((tv1.upper_bound is None and tv2.upper_bound is None)
+             or (tv1.upper_bound is not None
+                 and tv2.upper_bound is not None
+                 and is_same_type(tv1.upper_bound, tv2.upper_bound))))
 
 
 def is_same_arg_prefix(t: CallableType, s: CallableType) -> bool:

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -38,7 +38,7 @@ from mypy import erasetype
 from mypy.checkmember import analyze_member_access, type_object_type, bind_self
 from mypy.constraints import get_actual_type
 from mypy.checkstrformat import StringFormatterChecker
-from mypy.expandtype import expand_type, expand_type_by_instance
+from mypy.expandtype import expand_type, expand_type_by_instance, freshen_function_type_vars
 from mypy.util import split_module_names
 from mypy.semanal import fill_typevars
 from mypy.visitor import ExpressionVisitor
@@ -355,7 +355,7 @@ class ExpressionChecker(ExpressionVisitor[Type]):
                 lambda i: self.accept(args[i]))
 
             if callee.is_generic():
-                callee = freshen_generic_callable(callee)
+                callee = freshen_function_type_vars(callee)
                 callee = self.infer_function_type_arguments_using_context(
                     callee, context)
                 callee = self.infer_function_type_arguments(
@@ -2397,14 +2397,3 @@ def overload_arg_similarity(actual: Type, formal: Type) -> int:
         return 2
     # Fall back to a conservative equality check for the remaining kinds of type.
     return 2 if is_same_type(erasetype.erase_type(actual), erasetype.erase_type(formal)) else 0
-
-
-def freshen_generic_callable(callee: CallableType) -> CallableType:
-    tvdefs = []
-    tvmap = {}  # type: Dict[TypeVarId, Type]
-    for v in callee.variables:
-        tvdef = TypeVarDef.new_unification_variable(v)
-        tvdefs.append(tvdef)
-        tvmap[v.id] = TypeVarType(tvdef)
-
-    return cast(CallableType, expand_type(callee, tvmap)).copy_modified(variables=tvdefs)

--- a/mypy/checkmember.py
+++ b/mypy/checkmember.py
@@ -14,7 +14,7 @@ from mypy.nodes import (
 )
 from mypy.messages import MessageBuilder
 from mypy.maptype import map_instance_to_supertype
-from mypy.expandtype import expand_type_by_instance, expand_type
+from mypy.expandtype import expand_type_by_instance, expand_type, freshen_function_type_vars
 from mypy.infer import infer_type_arguments
 from mypy.semanal import fill_typevars
 from mypy import messages
@@ -80,6 +80,7 @@ def analyze_member_access(name: str,
             if is_lvalue:
                 msg.cant_assign_to_method(node)
             signature = function_type(method, builtin_type('builtins.function'))
+            signature = freshen_function_type_vars(signature)
             if name == '__new__':
                 # __new__ is special and behaves like a static method -- don't strip
                 # the first argument.

--- a/mypy/expandtype.py
+++ b/mypy/expandtype.py
@@ -1,13 +1,14 @@
-from typing import Dict, Iterable, List
+from typing import Dict, Iterable, List, TypeVar, Mapping, cast
 
 from mypy.types import (
     Type, Instance, CallableType, TypeVisitor, UnboundType, ErrorType, AnyType,
     Void, NoneTyp, TypeVarType, Overloaded, TupleType, TypedDictType, UnionType,
-    ErasedType, TypeList, PartialType, DeletedType, UninhabitedType, TypeType, TypeVarId
+    ErasedType, TypeList, PartialType, DeletedType, UninhabitedType, TypeType, TypeVarId,
+    FunctionLike, TypeVarDef
 )
 
 
-def expand_type(typ: Type, env: Dict[TypeVarId, Type]) -> Type:
+def expand_type(typ: Type, env: Mapping[TypeVarId, Type]) -> Type:
     """Substitute any type variable references in a type given by a type
     environment.
     """
@@ -28,12 +29,34 @@ def expand_type_by_instance(typ: Type, instance: Instance) -> Type:
         return expand_type(typ, variables)
 
 
+F = TypeVar('F', bound=FunctionLike)
+
+def freshen_function_type_vars(callee: F) -> F:
+    """Substitute fresh type variables for generic function type variables."""
+    if isinstance(callee, CallableType):
+        if not callee.is_generic():
+            return cast(F, callee)
+        tvdefs = []
+        tvmap = {}  # type: Dict[TypeVarId, Type]
+        for v in callee.variables:
+            tvdef = TypeVarDef.new_unification_variable(v)
+            tvdefs.append(tvdef)
+            tvmap[v.id] = TypeVarType(tvdef)
+        fresh = cast(CallableType, expand_type(callee, tvmap)).copy_modified(variables=tvdefs)
+        return cast(F, fresh)
+    else:
+        assert isinstance(callee, Overloaded)
+        fresh_overload = Overloaded([freshen_function_type_vars(item)
+                                     for item in callee.items()])
+        return cast(F, fresh_overload)
+
+
 class ExpandTypeVisitor(TypeVisitor[Type]):
     """Visitor that substitutes type variables with values."""
 
-    variables = None  # type: Dict[TypeVarId, Type]  # TypeVar id -> TypeVar value
+    variables = None  # type: Mapping[TypeVarId, Type]  # TypeVar id -> TypeVar value
 
-    def __init__(self, variables: Dict[TypeVarId, Type]) -> None:
+    def __init__(self, variables: Mapping[TypeVarId, Type]) -> None:
         self.variables = variables
 
     def visit_unbound_type(self, t: UnboundType) -> Type:

--- a/mypy/expandtype.py
+++ b/mypy/expandtype.py
@@ -31,6 +31,7 @@ def expand_type_by_instance(typ: Type, instance: Instance) -> Type:
 
 F = TypeVar('F', bound=FunctionLike)
 
+
 def freshen_function_type_vars(callee: F) -> F:
     """Substitute fresh type variables for generic function type variables."""
     if isinstance(callee, CallableType):

--- a/test-data/unit/check-generics.test
+++ b/test-data/unit/check-generics.test
@@ -1417,3 +1417,46 @@ class A:
     @classmethod
     def f(cls) -> None: pass
 [builtins fixtures/classmethod.pyi]
+
+[case testGenericOperatorMethodOverlapping]
+from typing import TypeVar, Generic, Tuple
+T = TypeVar('T')
+T2 = TypeVar('T2')
+S = TypeVar('S', bound=str)
+S2 = TypeVar('S2', bound=str)
+class G(Generic[T]):
+    pass
+class A:
+    def __or__(self, x: G[T]) -> G[T]: pass
+    def __ior__(self, x: G[T2]) -> G[T2]: pass
+class B:
+    def __or__(self, x: G[T]) -> G[T]: pass
+    def __ior__(self, x: G[S]) -> G[S]: pass \
+        # E: Signatures of "__ior__" and "__or__" are incompatible
+class C:
+    def __or__(self, x: G[S]) -> G[S]: pass
+    def __ior__(self, x: G[S2]) -> G[S2]: pass
+
+[case testGenericOperatorMethodOverlapping2]
+from typing import TypeVar, Generic, Tuple
+X = TypeVar('X')
+T = TypeVar('T', int, str)
+T2 = TypeVar('T2', int, str)
+S = TypeVar('S', float, str)
+S2 = TypeVar('S2', float, str)
+class G(Generic[X]):
+    pass
+class A:
+    def __or__(self, x: G[T]) -> G[T]: pass
+    def __ior__(self, x: G[T2]) -> G[T2]: pass
+class B:
+    def __or__(self, x: G[T]) -> G[T]: pass
+    def __ior__(self, x: G[S]) -> G[S]: pass \
+        # E: Signatures of "__ior__" and "__or__" are incompatible
+class C:
+    def __or__(self, x: G[S]) -> G[S]: pass
+    def __ior__(self, x: G[S2]) -> G[S2]: pass
+class D:
+    def __or__(self, x: G[X]) -> G[X]: pass
+    def __ior__(self, x: G[S2]) -> G[S2]: pass \
+        # E: Signatures of "__ior__" and "__or__" are incompatible

--- a/test-data/unit/check-inference-context.test
+++ b/test-data/unit/check-inference-context.test
@@ -865,3 +865,16 @@ reveal_type(c.f([1])) # E: Revealed type is 'builtins.list[builtins.int]'
 reveal_type(c.f([])) # E: Revealed type is 'builtins.list[builtins.int]'
 reveal_type(c.f(None)) # E: Revealed type is 'builtins.list[builtins.int]'
 [builtins fixtures/list.pyi]
+
+[case testGenericMethodCalledInGenericContext]
+from typing import TypeVar, Generic
+
+_KT = TypeVar('_KT')
+_VT = TypeVar('_VT')
+_T = TypeVar('_T')
+
+class M(Generic[_KT, _VT]):
+    def get(self, k: _KT, default: _T) -> _T: ...
+
+def f(d: M[_KT, _VT], k: _KT) -> _VT:
+    return d.get(k, None)

